src/program/lwaftr/tests/propbased/genyang.lua

```diff
@@ -0,0 +1,545 @@
+
+-- toggles whether functions should intentionally generate invalid
+-- values for fuzzing purposes
+local generate_invalid = true
+
+-- choose an element of an array randomly
+local function choose(choices)
+   local idx = math.random(#choices)
+   return choices[idx]
+end
+
+-- Generate a get/set/add/remove string given a pid string and optional schema
+function generate_any(pid, schema)
+   local cmd = choose({ "get", "add", "remove", "set" })
+
+   if cmd == "get" then
+      local query, schema = generate_config_xpath(schema)
+      return string.format("./snabb config get -s %s %s \"%s\"", schema, pid, query)
+   elseif cmd == "set" then
+      local query, val, schema = generate_config_xpath_and_val(schema)
+      return string.format("./snabb config set -s %s %s \"%s\" \"%s\"",
+                           schema, pid, query, val)
+   -- use rejection sampling for add and remove commands to restrict to list or
+   -- leaf-list cases (for remove, we need a case with a selector too)
+   -- Note: this assumes a list or leaf-list case exists in the schema at all
+   elseif cmd == "add" then
+      local query, val, schema
+      local ok = false
+      while not ok do
+         query, val, schema = generate_config_xpath_and_val(schema)
+         if string.match(tostring(val), "^{.*}$") then
+            ok = true
+         end
+      end
+      --local query, val, schema = generate_config_xpath_and_val(schema)
+      return string.format("./snabb config add -s %s %s \"%s\" \"%s\"",
+                           schema, pid, query, val)
+   else
+      local query, val, schema
+      local ok = false
+      while not ok do
+         query, val, schema = generate_config_xpath_and_val(schema)
+         if string.match(query, "[]]$") then
+            ok = true
+         end
+      end
+      return string.format("./snabb config remove -s %s %s \"%s\"",
+                           schema, pid, query)
+   end
+end
+
+-- Generate a get command string given a pid string and optional schema/query
+function generate_get(pid, schema, query)
+   if not query then
+      query, schema = generate_config_xpath(schema)
+   end
+   return string.format("./snabb config get -s %s %s \"%s\"", schema, pid, query)
+end
+
+-- Like generate_get but for state queries
+function generate_get_state(pid, schema, query)
+   if not query then
+      query, schema = generate_config_xpath_state(schema)
+   end
+   return string.format("./snabb config get-state -s %s %s \"%s\"", schema, pid, query)
+end
+
+-- Used primarily for repeating a set with a value seen before from a get
+function generate_set(pid, schema, query, val)
+   return string.format("./snabb config set -s %s %s \"%s\" \"%s\"",
+                        schema, pid, query, val)
+end
+
+function run_yang(yang_cmd)
+   local f = io.popen(yang_cmd)
+   local result = f:read("*a")
+   f:close()
+   return result
+end
+
+-- choose a natural number (e.g., index or length of array) by
+-- repeating a cointoss
```

Comment:
This is used often for the `[position=n]` part of a `leaf-list` query. Since most of the lists used in the config are small, it's more useful to test smaller numbers to access data that's actually in the config.

Though it's also used for, e.g., picking the length of a string. So another option would be to just bias, say, 1-5 and then pick a random number from a large range. I could implement that change if it seems better.